tests/queries/0_stateless/02353_explain_ast_rewrite.sql

```diff
@@ -0,0 +1,4 @@
+-- { echoOn }
+EXPLAIN AST rewrite=0 SELECT * FROM numbers(0);
```

Comment:
IMO would be nice to have tests with more interesting rewrites. e.g. `countDistinct` -> `uniqExact` and pushing where to subqueries

Reply:
It supposes to test that the setting has an effect, i.e. AST rewrite is happening. So, for this purpose, any simplest rewrite should be enough

Reply:
rewrite logic is not simple and users will expect that this command shows the actual result not just the initial query "after smth have happened". for this purpose we better have tests.

Reply:
Yes, I saw it. And this phase, hardly provide any guarantees. So, if we move some AST optimization to query plan level, it'll not be visible in this EXPLAIN any more.

Despite, I strive to name the setting `rewrite`. I also see that we're everywhere talking about optimization, like to in documentation to `EXPLAIN SYNTAX`. Should we rename the setting to `optimize` like proposed by @azat [here](https://github.com/ClickHouse/ClickHouse/pull/38910#discussion_r915043051) ? It feels more aligned to the context

P.S. "There are two hard things in computer science: cache invalidation, _naming_ _things_, and off-by-one errors" (c)

Reply:
I'll create follow-up PR for documentation

Reply:
See #38980 as follow-up PR

Reply:
>Should we rename the setting to optimize like proposed by @azat https://github.com/ClickHouse/ClickHouse/pull/38910#discussion_r915043051 ?

I like `optimize` more, since `EXPLAIN PLAN` has the same setting and this will make things simpler for user (although EXPLAIN are for experienced users)

By the way, maybe we should have the same setting for `EXPLAIN SYNTAX`, @devcrafter what do you think (but default behavior is different now, before `EXPLAIN SYNTAX` return optimized query, while `EXPLAIN AST` original, so either change the default to match but this is a minor backward compatibility or use different defaults which make people crazy) ?

Reply:
@azat `EXPLAIN SYNTAX` returns already both optimized and non-optimize syntaxes

Reply:
1. I like `optimize` a little bit more than `rewrite`. Both options are ok. Just, it is not so clear what we are going to rewrite in query. It is a little bit easier to say `we return the AST after optimizations`.
2. I still expect that EXPLAIN is more for development. I did not plan to document every setting and fix the behavior in docs. Current documentation to EXPLAIN gives a non-complete explanation what is returned, it does not give any guarantee. You are welcome to change it.
3. It is ok if we have setting `optimize` for both `EXPLAIN SYNTAX` (default `optimize = 1` ) and `EXPLAIN AST` (default `optimize = 0`).

Reply:
Ok, let's do it in follow-up #38980

Reply:
>@azat EXPLAIN SYNTAX returns already both optimized and non-optimize syntaxes

I don't see any settings for EXPLAIN SYNTAX.
https://github.com/ClickHouse/ClickHouse/blob/d0e6f8ebc120a682bb09231fbbba1a904c8c29bb/src/Interpreters/InterpreterExplainQuery.cpp#L297

Or you mean that the original query will be printed anyway? Hence there is no point in providing such setting?
Then I guess you are right.

Reply:
> Or you mean that the original query will be printed anyway? Hence there is no point in providing such setting?
Then I guess you are right.

Yes, that's what I meant